File: application/src/types/protocol.ts
// FrostfireOS IPC Protocol Types
// Must match field-for-field with src-tauri/src/protocol.rs in snake_case

export interface ActionItem {
  id: string;
  description: string;
  assigned_agent_id: string;
  priority: string;
}

export interface RiskEvaluation {
  risk: string;
  severity: string;
  mitigation: string;
}

export interface TeamPlan {
  workstream_id: string;
  team_id: string;
  title: string;
  executive_summary: string;
  action_items: ActionItem[];
  target_files: string[];
  execution_order: string[];
  identified_risks: RiskEvaluation[];
}

export interface ToolApprovalRequest {
  request_id: string;
  agent_id: string;
  tool_name: string;
  parameters: Record<string, unknown>;
  rationale: string;
}

export interface BlackboardArtifact {
  uri: string;
  author_id: string;
  title: string;
  content: string;
  mime_type: string;
  version: number;
  hash: string;
  created_at: string;
}

export type WorkstreamCommand =
  | {
      command: "launch_workstream";
      blueprint_id: string;
      workstream_name: string;
      params: Record<string, unknown>;
    }
  | {
      command: "pause_workstream";
      workstream_id: string;
    }
  | {
      command: "approve_milestone";
      workstream_id: string;
      milestone_id: string;
    }
  | {
      command: "read_blackboard";
      uri: string;
    }
  | {
      command: "save_api_key";
      provider: string;
      key: string;
    }
  | {
      command: "get_system_status";
    }
  | {
      command: "calibrate_fta";
      workstream_id: string;
      rating: number;
      hours_saved: number;
    }
  | {
      command: "get_prompt_config";
      role: string;
    }
  | {
      command: "save_custom_prompt";
      role: string;
      content: string;
      activate: boolean;
    }
  | {
      command: "get_blackboard_manifest";
      board_id: string;
    }
  | {
      command: "get_blackboard_presentation";
      board_id: string;
    }
  | {
      command: "verify_invariants";
      board_id: string;
    }
  | {
      command: "web_search";
      query: string;
    };

export interface SearchResult {
  title: string;
  snippet: string;
  url: string;
}

export interface PromptConfigDto {
  role: string;
  is_custom: boolean;
  display_status: string;
  prompt_content: string;
}

export type WorkstreamEvent =
  | {
      event: "sme_task_started";
      task_id: string;
      agent_id: string;
      role: string;
      phase: string;
    }
  | {
      event: "artifact_published";
      uri: string;
      title: string;
      author: string;
      version: number;
      size_bytes: number;
    }
  | {
      event: "team_plan_synthesized";
      workstream_id: string;
      plan: TeamPlan;
    }
  | {
      event: "tool_approval_request";
      request_id: string;
      agent_id: string;
      tool_name: string;
      parameters: Record<string, unknown>;
      rationale: string;
    }
  | {
      event: "workstream_completed";
      workstream_id: string;
      total_hours_saved: number;
    }
  | {
      event: "token_stream";
      task_id: string;
      agent_id: string;
      chunk: string;
    };

export interface SystemStatus {
  version: string;
  os: string;
  app_data_dir: string;
  extensions_dir: string;
  connected_providers: string[];
  active_workstreams_count: number;
  total_labor_hours_saved: number;
}

export interface TestKeyResponse {
  success: boolean;
  latency_ms: number;
  message: string;
  models?: string[];
}

File: application/src-tauri/src/protocol.rs
use frostfire_core::synthesis::TeamPlan;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "command", rename_all = "snake_case")]
pub enum WorkstreamCommand {
    LaunchWorkstream {
        blueprint_id: String,
        workstream_name: String,
        params: serde_json::Value,
    },
    PauseWorkstream {
        workstream_id: String,
    },
    ApproveMilestone {
        workstream_id: String,
        milestone_id: String,
    },
    ReadBlackboard {
        uri: String,
    },
    SaveApiKey {
        provider: String,
        key: String,
    },
    GetSystemStatus,
    CalibrateFta {
        workstream_id: String,
        rating: u8,
        hours_saved: f64,
    },
    GetPromptConfig {
        role: String,
    },
    SaveCustomPrompt {
        role: String,
        content: String,
        activate: bool,
    },
    GetBlackboardManifest {
        board_id: String,
    },
    GetBlackboardPresentation {
        board_id: String,
    },
    VerifyInvariants {
        board_id: String,
    },
    WebSearch {
        query: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult {
    pub title: String,
    pub snippet: String,
    pub url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PromptConfigDto {
    pub role: String,
    pub is_custom: bool,
    pub display_status: String,
    pub prompt_content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", rename_all = "snake_case")]
pub enum WorkstreamEvent {
    SmeTaskStarted {
        task_id: String,
        agent_id: String,
        role: String,
        phase: String,
    },
    ArtifactPublished {
        uri: String,
        title: String,
        author: String,
        version: u32,
        size_bytes: usize,
    },
    TeamPlanSynthesized {
        workstream_id: String,
        plan: TeamPlan,
    },
    ToolApprovalRequest {
        request_id: String,
        agent_id: String,
        tool_name: String,
        parameters: serde_json::Value,
        rationale: String,
    },
    WorkstreamCompleted {
        workstream_id: String,
        total_hours_saved: f64,
    },
    TokenStream {
        task_id: String,
        agent_id: String,
        chunk: String,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SystemStatus {
    pub version: String,
    pub os: String,
    pub app_data_dir: String,
    pub extensions_dir: String,
    pub connected_providers: Vec<String>,
    pub active_workstreams_count: usize,
    pub total_labor_hours_saved: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TestKeyResponse {
    pub success: bool,
    pub latency_ms: u64,
    pub message: String,
    #[serde(default)]
    pub models: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum RpcEvent {
    SessionStart {
        session_id: String,
        model: String,
        provider: String,
    },
    TurnStart {
        turn_number: usize,
        prompt: String,
    },
    TextChunk {
        content: String,
    },
    ReasoningChunk {
        content: String,
    },
    TurnEnd {
        turn_number: usize,
    },
    Error {
        code: String,
        message: String,
    },
}
